### ISSUE TITLE
LLM Hyperparameters (dim, heads, layers, ff_dim, batch_size) Search

### ISSUE THREAD

**Issue by vukrosic** [Open-Superintelligence-Lab/5-dollar-llm#58]
Searching for the best LLM model hyperparams.

I like these 2 because the first one is very small and trains fast, and the second one is big enough to be smarter but still fast to train.

```bash
         Architecture  Params (M)  Max VRAM (GB)  Target Loss  Final Loss  Steps  Training Time (s)  Total Time (s)
    BlueberryB8Config        96.0          21.28          3.5    4.285505   1752         344.587858      357.776236
BlueberryB8WideConfig       163.6          21.15          3.5    4.246539   1752         425.077524      438.308722
```

<img width="3086" height="2113" alt="Image" src="https://github.com/user-attachments/assets/76001643-e012-4026-aae5-fe270af96a30" />

### Full search:

```
##################################################
📊 BENCHMARK SUMMARY (Target Loss: 3.5)
##################################################
         Architecture  Params (M)  Max VRAM (GB)  Target Loss  Final Loss  Steps  Training Time (s)  Total Time (s)
      BlueberryConfig       151.0          14.49          3.5    4.190339   2140         389.188305      403.781130
  BlueberryWideConfig       250.1          18.13          3.5    4.155141   2140         531.465466      549.120020
   BlueberryFFNConfig       226.5          17.85          3.5    4.186517   2140         478.186726      494.925372
    BlueberryB8Config        96.0          21.28          3.5    4.285505   1752         344.587858      357.776236
BlueberryB8WideConfig       163.6          21.15          3.5    4.246539   1752         425.077524      438.308722
BlueberryB8TallConfig        46.2          20.74          3.5    4.427349   1752         333.985906      348.102440

Summary saved to benchmarks/summary_loss3.5.csv
root@df4ce776f0874983:~/5-dollar-llm# 
```

### Code for the configs:

```python

@dataclass
class BlueberryConfig:
    # Model architecture (151M Params - Blueberry-Nano)
    d_model: int = 512       
    n_heads: int = 8         
    n_layers: int = 32       
    d_ff: int = 2048         
    
    # GQA parameters
    n_kv_heads: int = 4      
    
    # Data params
    max_seq_len: int = 2048  
    vocab_size: int = 49152  
    
    # Base Training Defaults
    compile_model: bool = True
    batch_size: int = 4
    gradient_accumulation_steps: int = 1
    train_tokens: int = 20000000
    
    # Learning Rate (Aggressive for pre-training)
    muon_lr: float = 0.015
    muon_momentum: float = 0.95
    adamw_lr: float = 0.001
    warmup_ratio: float = 0.0
    schedule_type: str = "constant"

    # Evaluation
    eval_every: int = 2000
    eval_steps: int = 100
    
    # Regularization
    weight_decay: float = 0.2
    dropout: float = 0.0
    grad_clip: float = 1.0
    use_amp: bool = True
    
    # Logging
    log_milestones: Tuple[int, ...] = (100, 500, 1000)

    def __post_init__(self):
        self.d_k = self.d_model // self.n_heads
        assert self.d_model % self.n_heads == 0, "d_model must be divisible by n_heads"


@dataclass
class BlueberryTallConfig(BlueberryConfig):
    # ~160M Params - Deeper and thinner
    d_model: int = 384
    n_heads: int = 6
    n_layers: int = 64
    d_ff: int = 1536
    n_kv_heads: int = 3
    # Best guess: Conservative LR for high depth
    muon_lr: float = 0.012
    adamw_lr: float = 0.0008


@dataclass
class BlueberryWideConfig(BlueberryConfig):
    # ~250M Params - Shallower and wider
    d_model: int = 768
    n_heads: int = 12
    n_layers: int = 30
    d_ff: int = 2304
    n_kv_heads: int = 6
    # Best guess: Standard LR
    muon_lr: float = 0.015


@dataclass
class BlueberryFFNConfig(BlueberryConfig):
    # ~226M Params - High FFN-to-d_model ratio
    d_model: int = 512
    n_heads: int = 8
    n_layers: int = 32
    d_ff: int = 3584
    n_kv_heads: int = 4
    # Best guess: Standard LR
    muon_lr: float = 0.015


@dataclass
class BlueberryB8Config(BlueberryConfig):
    # Optimized for Batch 8
    batch_size: int = 8
    d_model: int = 512
    n_heads: int = 8
    n_layers: int = 18
    d_ff: int = 2048
    n_kv_heads: int = 4
    # Best guess: Scaled LR for double batch
    muon_lr: float = 0.02
    adamw_lr: float = 0.0012


@dataclass
class BlueberryB8WideConfig(BlueberryConfig):
    # Optimized for Batch 8 - Wider model
    batch_size: int = 8
    d_model: int = 1024
    n_heads: int = 16
    n_layers: int = 9
    d_ff: int = 3072
    n_kv_heads: int = 8
    # Best guess: Scaled LR for double batch
    muon_lr: float = 0.02
    adamw_lr: float = 0.0012


@dataclass
class BlueberryB8TallConfig(BlueberryConfig):
    # Optimized for Batch 8 - Deeper model
    batch_size: int = 8
    d_model: int = 256
    n_heads: int = 4
    n_layers: int = 32
    d_ff: int = 1024
    n_kv_heads: int = 4
    # Best guess: Scaled LR for double batch
    muon_lr: float = 0.02
    adamw_lr: float = 0.0012
```
